### PR TITLE
add LLVM 4.0.1

### PR DESCRIPTION
**Pull request by dvirtz**

### CODE REVIEW

**Comment by dvirtz** [comment]
resolves https://github.com/ruslo/hunter/issues/918.

I didn't update Libcxx as it isn't used on any example and I don't have a way to test it

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2115
* https://travis-ci.org/ingenue/hunter/builds/262126222?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
```
CMake Error at cmake/modules/CheckCompilerVersion.cmake:45 (message):
  Host Visual Studio must be at least 2015
```

```
CMake Error at cmake/modules/AddLLVM.cmake:1340 (string):
  string begin index: -1 is out of range 0 - 23
```

Please exclude all toolchain that failed not by timeout. For timeout I will upload binaries from local machine.

**Comment by dvirtz** [comment]
I don't see any timeout, do you?

```
CMake Error at cmake/modules/AddLLVM.cmake:1340 (string):
  string begin index: -1 is out of range 0 - 23
```

This happens because Hunter passes `-DCMAKE_CONFIGURATION_TYPES=${configuration} ` even on single configuration generators and LLVM expects it being set only on multi-configuration generators.
What do you think about changing the CMake scheme to do
```cmake
  if(CMAKE_CONFIGURATION_TYPES)
    set(build_type_opts "-DCMAKE_CONFIGURATION_TYPES=${configuration}")
  else()
    set(build_type_opts "-DCMAKE_BUILD_TYPE=${configuration}")
  endif()
```
and then passing only `build_type_opts` to CMake?

**Comment by ruslo** [comment]
> I don't see any timeout, do you?

I saw that `osx-10-11` takes 48 minutes, but now I'm not sure about reason of failure.

> What do you think about changing the CMake scheme to do

We can use only `CMAKE_CONFIGURATION_TYPES` on multi and only `CMAKE_BUILD_TYPE` on single. But check [should be](http://cgold.readthedocs.io/en/latest/tutorials/build-types.html#detect-multi-single):
```cmake
string(COMPARE EQUAL "${CMAKE_CFG_INTDIR}" "." is_single)
if(is_single)
  set(build_type_opts "-DCMAKE_BUILD_TYPE=${configuration}")
else()
  set(build_type_opts "-DCMAKE_CONFIGURATION_TYPES=${configuration}")
endif()
```

> LLVM expects it being set only on multi-configuration generators

LLVM should not expect this. It's a separate issue and should be fixed in LLVM.

**Comment by ruslo** [comment]
> set build type differently on single and multi configuration generators

Please open separate pull request with this change. Want to run few tests.

**Comment by dvirtz** [comment]
Done. https://github.com/ruslo/hunter/pull/948

**Comment by ruslo** [comment]
> Done. #948

Merged, thanks. Please rebase this one.

File: cmake/projects/PocoCpp/hunter.cmake [hunk 1]
```diff
@@ -8,6 +8,17 @@ hunter_add_version(
     PACKAGE_NAME
     PocoCpp
     VERSION
+    1.7.8-p0
+    URL
+    "https://github.com/hunter-packages/poco/archive/v1.7.8-p0.zip"
```
Comment (ruslo) [inline cmake/projects/PocoCpp/hunter.cmake]: rebase + squash?

**Comment by dvirtz** [comment]
sorry, it always takes me some tries to make it right :-)

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2126
* https://travis-ci.org/ingenue/hunter/builds/262791336?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
Tests failed. I see timeout, the rest should be removed/fixed.

**Comment by ruslo** [comment]
New toolchain added, retest:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2131
* https://travis-ci.org/ingenue/hunter/builds/262829881?utm_source=github_status&utm_medium=notification

**Comment by dvirtz** [comment]
OK, VS 2015 now fails because of long paths. Can I modify AppVeyor receipe to use `HUNTER_BINARY_DIR`?

**Comment by ruslo** [comment]
> OK, VS 2015 now fails because of long paths. Can I modify AppVeyor receipe to use HUNTER_BINARY_DIR?

No need, I want to upload it from local machine anyway.

**Comment by ruslo** [comment]
Toolchain `osx-10-11` failed on my local machine with error:
```
=== BUILD TARGET clang_rt.builtins_x86_64h_osx OF PROJECT LLVM WITH CONFIGURATION Release ===

Check dependencies
No architectures to compile for (ARCHS=x86_64h, VALID_ARCHS=i386 x86_64).

** BUILD FAILED **
```

Don't ask me what it means, I have no idea :) I guess should be excluded too.

Working so far:
* `vs-14-2015`: https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2135
* `libcxx`: https://travis-ci.org/ingenue/hunter/jobs/262829885

**Comment by ruslo** [comment]
> I guess should be excluded

`default` is not very interesting because it's quite the same as `libcxx` so I vote for excluding too. It will save some time on local upload. Same for Linux `default` ~ `gcc`.

**Comment by ruslo** [comment]
`gcc` is working now:
* https://travis-ci.org/ingenue/hunter/jobs/262829883

We have `vs-14-2015` on Windows, `libcxx` on OSX and `gcc` on Linux. Please exclude the rest.

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2144
* https://travis-ci.org/ingenue/hunter/builds/263496307?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
Released: https://github.com/ruslo/hunter/releases/tag/v0.19.68

**Pull request merged**